sycl-jit/jit-compiler/lib/rtc/RTC.cpp

```diff
@@ -0,0 +1,168 @@
+    TraceFileName = Arg->getValue();
+    int Granularity =
+        500; // microseconds. Same default as in `clang::FrontendOptions`.
+    if (auto *Arg = UserArgList.getLastArg(
+            clang::driver::options::OPT_ftime_trace_granularity_EQ)) {
+      if (!llvm::to_integer(Arg->getValue(), Granularity)) {
+        BuildLog += "warning: ignoring malformed argument: '" +
+                    Arg->getAsString(UserArgList) + "'\n";
+      }
+    }
+    bool Verbose =
+        UserArgList.hasArg(clang::driver::options::OPT_ftime_trace_verbose);
+
+    llvm::timeTraceProfilerInitialize(Granularity, /*ProcName=*/"sycl-rtc",
+                                      Verbose);
+  }
+
+  std::unique_ptr<llvm::Module> Module;
+
+  if (CachedIR.size() > 0) {
+    llvm::StringRef IRStr{CachedIR.begin(), CachedIR.size()};
+    std::unique_ptr<llvm::MemoryBuffer> IRBuf =
+        llvm::MemoryBuffer::getMemBuffer(IRStr, /*BufferName=*/"",
+                                         /*RequiresNullTerminator=*/false);
+    auto ModuleOrError = llvm::parseBitcodeFile(*IRBuf, Context);
+    if (!ModuleOrError) {
+      // Not a fatal error, we'll just compile the source string normally.
+      BuildLog.append(formatError(ModuleOrError.takeError(),
+                                  "Loading of cached device code failed"));
+    } else {
+      Module = std::move(*ModuleOrError);
+    }
+  }
+
+  bool FromSource = false;
+  if (!Module) {
+    auto ModuleOrErr = compileDeviceCode(SourceFile, IncludeFiles, UserArgList,
+                                         BuildLog, Context);
+    if (!ModuleOrErr) {
+      return errorTo<RTCResult>(ModuleOrErr.takeError(),
+                                "Device compilation failed");
+    }
+
+    Module = std::move(*ModuleOrErr);
+    FromSource = true;
+  }
+
+  RTCDeviceCodeIR IR;
+  if (SaveIR && FromSource) {
+    std::string BCString;
+    llvm::raw_string_ostream BCStream{BCString};
+    llvm::WriteBitcodeToFile(*Module, BCStream);
+    IR = RTCDeviceCodeIR{BCString.data(), BCString.data() + BCString.size()};
+  }
+
+  if (auto Error = linkDeviceLibraries(*Module, UserArgList, BuildLog)) {
+    return errorTo<RTCResult>(std::move(Error), "Device linking failed");
+  }
+
+  auto PostLinkResultOrError = performPostLink(std::move(Module), UserArgList);
+  if (!PostLinkResultOrError) {
+    return errorTo<RTCResult>(PostLinkResultOrError.takeError(),
+                              "Post-link phase failed");
+  }
+  auto [BundleInfo, Modules] = std::move(*PostLinkResultOrError);
+
+  for (auto [DevImgInfo, Module] :
+       llvm::zip_equal(BundleInfo.DevImgInfos, Modules)) {
+    auto BinaryInfoOrError = Translator::translate(
+        *Module, JITContext::getInstance(), BinaryFormat::SPIRV);
+    if (!BinaryInfoOrError) {
+      return errorTo<RTCResult>(BinaryInfoOrError.takeError(),
+                                "SPIR-V translation failed");
+    }
+    DevImgInfo.BinaryInfo = std::move(*BinaryInfoOrError);
```

Comment:
Nit; I like `Expected::moveInto`, but I won't force you to use it.
```suggestion
    if (auto Error = Translator::translate(*Module, JITContext::getInstance(),
                                           BinaryFormat::SPIRV)
                         .moveInto(DevImgInfo.BinaryInfo)) {
      return errorTo<RTCResult>(std::move(Error), "SPIR-V translation failed");
    }
```